File: internal/models/tasklist.go
package models

import "time"

// TasklistItem represents a checklist item within an issue
type TasklistItem struct {
	ID          int        `json:"id"`
	IssueID     int        `json:"issue_id"`
	Content     string     `json:"content"`
	IsCompleted bool       `json:"is_completed"`
	Position    int        `json:"position"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
	CompletedBy *int       `json:"completed_by,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`

	// Populated fields
	CompletedByUser *User `json:"completed_by_user,omitempty"`
}

// TasklistProgress represents the completion progress of a tasklist
type TasklistProgress struct {
	Total     int `json:"total"`
	Completed int `json:"completed"`
	Pending   int `json:"pending"`
	Percent   int `json:"percent"`
}

// CreateTasklistItemRequest represents a request to create a tasklist item
type CreateTasklistItemRequest struct {
	Content  string `json:"content" validate:"required,min=1,max=500"`
	Position *int   `json:"position,omitempty"`
}

// UpdateTasklistItemRequest represents a request to update a tasklist item
type UpdateTasklistItemRequest struct {
	Content     *string `json:"content,omitempty" validate:"omitempty,min=1,max=500"`
	IsCompleted *bool   `json:"is_completed,omitempty"`
	Position    *int    `json:"position,omitempty"`
}

// ReorderTasklistRequest represents a request to reorder tasklist items
type ReorderTasklistRequest struct {
	ItemIDs []int `json:"item_ids" validate:"required,min=1"`
}

// BulkCreateTasklistRequest represents a request to create multiple tasklist items
type BulkCreateTasklistRequest struct {
	Items []CreateTasklistItemRequest `json:"items" validate:"required,min=1,max=50,dive"`
}

File: frontend/src/api/tasklist.ts
import { apiClient } from './client';
import type {
  TasklistItem,
  TasklistProgress,
  CreateTasklistItemRequest,
  UpdateTasklistItemRequest,
  ReorderTasklistRequest,
  BulkCreateTasklistRequest,
} from '../types';

// Get all tasklist items for an issue
export const getTasklistItems = async (issueId: number): Promise<TasklistItem[]> => {
  const response = await apiClient.get<TasklistItem[]>(`/issues/${issueId}/tasklist`);
  return response.data;
};

// Get tasklist progress for an issue
export const getTasklistProgress = async (issueId: number): Promise<TasklistProgress> => {
  const response = await apiClient.get<TasklistProgress>(`/issues/${issueId}/tasklist/progress`);
  return response.data;
};

// Create a new tasklist item
export const createTasklistItem = async (
  issueId: number,
  data: CreateTasklistItemRequest
): Promise<TasklistItem> => {
  const response = await apiClient.post<TasklistItem>(`/issues/${issueId}/tasklist`, data);
  return response.data;
};

// Bulk create tasklist items
export const bulkCreateTasklistItems = async (
  issueId: number,
  data: BulkCreateTasklistRequest
): Promise<TasklistItem[]> => {
  const response = await apiClient.post<TasklistItem[]>(`/issues/${issueId}/tasklist/bulk`, data);
  return response.data;
};

// Update a tasklist item
export const updateTasklistItem = async (
  itemId: number,
  data: UpdateTasklistItemRequest
): Promise<TasklistItem> => {
  const response = await apiClient.put<TasklistItem>(`/tasklist/${itemId}`, data);
  return response.data;
};

// Toggle tasklist item completion
export const toggleTasklistItem = async (itemId: number): Promise<TasklistItem> => {
  const response = await apiClient.patch<TasklistItem>(`/tasklist/${itemId}/toggle`);
  return response.data;
};

// Delete a tasklist item
export const deleteTasklistItem = async (itemId: number): Promise<void> => {
  await apiClient.delete(`/tasklist/${itemId}`);
};

// Reorder tasklist items
export const reorderTasklistItems = async (
  issueId: number,
  data: ReorderTasklistRequest
): Promise<void> => {
  await apiClient.put(`/issues/${issueId}/tasklist/reorder`, data);
};
